hm_14/hm_014.py

```diff
@@ -0,0 +1,82 @@
+import re
+
+# task_num_1
+phone_book = {
+    "Alina Kravchenko": "0664686501",
+    "Egor Miroshnichenko": "0995678456",
+    "Ruslan Volikov": "0987656342",
+    "Nastya Lehenka": "0995676456"
+
+}
+count = 3
+while count:
+    user_input = input("Enter a command here: ")
+    split_input = user_input.split()
+
+    if user_input == "stats":
+        contacts_amount = len(phone_book)
+        print(contacts_amount)
+
+    elif user_input == "list":
+        all_names = phone_book.keys()
+        for name in all_names:
+            print(name)
+
+    elif "show" in user_input:  # format: "show full name"
+        name = split_input[1] + " " + split_input[2]
+        print(phone_book.get(name.title()))
+
+    elif "add" in user_input:  # format: "add full name number"
+        new_name = split_input[1].title() + " " + split_input[2].title()
+        if phone_book.get(new_name) is None:
+            number_val = re.search("[\+380|80|0]\d{9}", split_input[3])
```

Comment:
Your regular expression allows almost all invalid formats of the phone number. Here are some test cases and the result:
```
a0951234567 -> 0951234567
8+30951234567 -> +309512345
a+380951234567 -> +380951234
+380951234567a -> +380951234
+380951234567a -> +380951234
+3809512345677 -> +380951234
80951234567 -> 8095123456
+830965254141 -> +830965254
```
To improve it:
- use fullmatch instead of search, it will fix all invalid cases
- update your pattern: 80XXXXXXXXX is invalid format, use 380XXXXXXXXX instead
- check your solution for finding a valid formats

Reply:
It looks better now, but it allow only one valid format: 0XXXXXXXXX
I can suggest using of this pattern: `(\+)?(38)?0\d{9}`
Here we group separate parts of separate formats and by adding `?` point that these parts are optional